lib/booking_system.rb

```diff
@@ -0,0 +1,149 @@
+module Hotel
+
+  class BookingSystem
+
+    attr_reader :all_rooms, :all_reservations, :all_room_blocks
+
+    def initialize(all_reservations: [])
+      @all_rooms = [*1..20]
+      @all_reservations = all_reservations
+      @all_room_blocks = []
+    end
+
+
+    def list_rooms
+      return @all_rooms.map { |room_num| "Room #{room_num}"}.join("\n")
+    end
+
+
+    def list_printable_available_rooms(check_in, check_out)
+      range = Hotel::DateRange.new(check_in, check_out)
+      room_list = list_available_rooms(range)
+      return "Available Rooms:\n" + room_list.map { |room_num| "Room #{room_num}"}.join("\n")
+    end
+
+
+    def list_available_rooms(range)
+        occupied_rooms = list_unavailable_rooms(range)
+        return @all_rooms - occupied_rooms
+    end
+
+
+    def list_unavailable_rooms(range)
+      overlapping_reservations = @all_reservations.select { |reservation| reservation.date_range.overlaps?(range) }
+
+      occupied_rooms = overlapping_reservations.map { |reservation| reservation.room_num}
+
+      overlapping_blocks = @all_room_blocks.select { |block|
+        block.date_range.overlaps?(range) }
+
+      overlapping_blocks.each do |block|
+        occupied_rooms.concat block.blocked_rooms
+      end
+
+      return occupied_rooms
+    end
+
+
+    def list_reservations_by_date(date)
+      return @all_reservations.select do |reservation|
+        reservation.date_range.is_within_date_range(date)
+      end
+    end
+
+
+    def make_reservation(range, room, block: false)
+      unless block
+
+        if !is_room_available(range, room)
+          raise UnavailableRoomError, "Room not available"
+        end
+
+        new_reservation = Hotel::Reservation.new(range, room)
+        @all_reservations << new_reservation
+
+      else
+
+        if block.blocked_rooms.include?(room)
+          new_reservation = Hotel::Reservation.new(range, room, rate: block.discounted_rate)
+          block.block_reservations << new_reservation
+        else
+          raise UnavailableRoomError, "Room in different block"
+        end
+
+      end
+
+      return new_reservation
+    end
+
+
+    def reserve_standard_room(check_in, check_out)
+      range = Hotel::DateRange.new(check_in, check_out)
+
+      open_rooms = list_available_rooms(range)
+
+      if open_rooms.empty?
+        raise UnavailableRoomError, "No Rooms available for #{check_in} to #{check_out}"
+      else
+        make_reservation(range, open_rooms.first)
+      end
+    end
+
+
+    def reserve_a_room_in_block(id)
+      block = find_block(id)
+
+      unless block
+        raise ArgumentError, "Block not found in system"
+      end
+
+      unless block.get_block_availability
+        raise ArgumentError, "Block has no availability"
+      end
+
+      room_num = block.list_available_block_rooms.first
+      make_reservation(block.date_range, room_num, block: block)
```

Comment:
I LOVE THIS. You used the optional/named keyword argument and it was super useful in this case! Wonderful.